userdocs/diagnostics/trailing-closure-matching.md

```diff
@@ -0,0 +1,99 @@
+
+```swift
+// Remains valid
+animate(withDuration: 0.3, animations: { /* Animate Something */ }) {
+  // Done!
+}
+
+// Also OK!
+animate(withDuration: 0.3) {
+  // Animate Something
+}
+```
+
+When scanning forwards to match an unlabeled trailing closure argument, the compiler may sometimes need to "skip over" defaulted and variadic arguments. The new rule will skip any parameter that does not structurally resemble a function type. This allows writing a modified version of the above example where `withDuration` also has a default argument value:
+
+```swift
+func animate(
+  withDuration duration: Double = 1.0, 
+  animations: () -> Void, 
+  completion: (() -> Void)? = nil
+) {}
+
+// Allowed! The forward scanning rule skips `withDuration` because it does not structurally resemble a function type.
+animate {
+  // Animate Something
+}
+```
+
+A parameter structurally resembles a function type if both of the following are true:
+
+- The parameter is not `inout`
+- The adjusted type of the parameter is a function type
+
+The adjusted type of the parameter is the parameter's type as it appears in the function declaration, looking through any type aliases, and performing three additional adjustments:
+
+- If the parameter is an `@autoclosure`, using the result type of the parameter's declared (function) type, before performing the second adjustment.
+- If the parameter is variadic, looking at the base element type.
+- Removing all outer "optional" types.
+
+To maintain source compatibility with code that was written before Swift 5.3, the forward scanning rule applies an additional heuristic when matching trailing closure arguments. If,
+
+- the parameter that would match an unlabeled trailing closure argument according to the forward scanning rule does not require an argument (because it is variadic or has a default argument), _and_
+- there are parameters _following_ that parameter that _do_ require an argument, which appear before the first parameter whose label matches that of the _next_ trailing closure (if any)
+
+then the compiler does not match the unlabeled trailing closure to that parameter. Instead, it skips it and examines the next parameter to see if that should be matched against the unlabeled trailing closure. This can be seen in the following example:
+
+```swift
+func showAlert(
+  message: String,
+  onPresentation: (() ->  Void)? = nil,
+  onDismissal: () -> Void
+) {}
+
+// The unlabeled trailing closure matches `onDismissal` because `onPresentation` does not require an argument,
+// but `onDismissal` does and there are no other trailing closures which could match it.
```

Comment:
Wrap to 80 characters?

```suggestion
// The unlabeled trailing closure matches `onDismissal` because `onPresentation`
// does not require an argument, but `onDismissal` does and there are no other
// trailing closures which could match it.
```